pkg/apis/core/types.go

```diff
@@ -3510,17 +3510,39 @@ type LoadBalancerIngress struct {
 	Hostname string
 }
 
+const (
+	// MaxServiceTopologyKeys is the largest number of topology keys allowed on a service
+	MaxServiceTopologyKeys = 16
+)
+
 // IPFamily represents the IP Family (IPv4 or IPv6). This type is used
-// to express the family of an IP expressed by a type (i.e. service.Spec.IPFamily)
+// to express the family of an IP expressed by a type (i.e. Service.Spec.IPFamilies).
 type IPFamily string
 
+// IPFamilyPolicyType represents the dualstack-ness requested or required by a Service
+type IPFamilyPolicyType string
+
 const (
 	// IPv4Protocol indicates that this IP is IPv4 protocol
 	IPv4Protocol IPFamily = "IPv4"
 	// IPv6Protocol indicates that this IP is IPv6 protocol
 	IPv6Protocol IPFamily = "IPv6"
-	// MaxServiceTopologyKeys is the largest number of topology keys allowed on a service
-	MaxServiceTopologyKeys = 16
+	// SingleStack indicates that this service is required to have a single IPFamily
+	// The IPFamily assigned is based on the default IPFamily used by the cluster
+	// or as identified by Spec.IPFamilies field
+	SingleStack IPFamilyPolicyType = "SingleStack"
+	// PreferDualStack indicates that this service prefers dualstack (two IPFamilies) whenever
+	// the cluster is configured for dual stack. If the cluster is not configured
+	// with dualstack the service will be assigned single IPFamily. If the IPFamily is not
+	// set in Service.Spec.IPFamilies then the service will be assigned the default IPFamily
+	// configured on the cluster
+	PreferDualStack IPFamilyPolicyType = "PreferDualStack"
+	// RequireDualStack indicate that this server requires dualstack (two IPFamilies). Using
+	// RequireDualStack on a single stack cluster will result in failure. The IPFamilies (and their order) assigned
+	// to this service is based on Service.Spec.IPFamilies. If Service.Spec.IPFamilies was not provided
+	// then IPFamilies will be assigned according to how they are configured on the cluster. If
+	// Service.Spec.IPFamilies has only one entry then the alternative IPFamily will be added by apiserver
```

Comment:
nit: "then other IPFamily will be appended to the list. For example, if Service.Spec.IPFamilies = [IPv6Protocol], then the resulting IPFamilies will be [IPv6Protocol, IPv4Protocol]"